File: backend/app/routers/groups.py
from uuid import UUID

from fastapi import APIRouter, Depends

from app.models.groups import (
    AddMemberRequest,
    GroupCreate,
    GroupDetailOut,
    GroupMemberOut,
    GroupOut,
)
from app.models.pagination import PaginatedResponse
from app.models.users import UserOut
from app.services.auth_service import get_current_user
from app.services.group_service import GroupService

router = APIRouter(prefix="/groups", tags=["Groups"])

group_service = GroupService()


def get_group_service() -> GroupService:
    return group_service


@router.post("/", response_model=GroupOut)
async def create_group(
    data: GroupCreate,
    current_user: UserOut = Depends(get_current_user),
    service: GroupService = Depends(get_group_service),
):
    return await service.create_group(data, current_user.id)


@router.get("/", response_model=PaginatedResponse[GroupOut])
async def get_user_groups(
    search: str | None = None,
    filter: str | None = None,
    sort_by: str = "created_at",
    order: str = "desc",
    skip: int = 0,
    limit: int = 20,
    current_user: UserOut = Depends(get_current_user),
    service: GroupService = Depends(get_group_service),
):

    """
    Get all groups for the current user with search, filter, sort, and pagination.
    """
    return await service.get_user_groups(
        user_id=current_user.id,
        search=search,
        filter=filter,
        sort_by=sort_by,
        order=order,
        skip=skip,
        limit=limit,
    )




@router.get("/{group_id}", response_model=GroupDetailOut)
async def get_group(
    group_id: UUID,
    current_user: UserOut = Depends(get_current_user),
    service: GroupService = Depends(get_group_service),
):
    """
    Get group details (members only).
    """
    return await service.get_group_detail(str(group_id), current_user.id)


@router.post("/{group_id}/members", response_model=GroupMemberOut)
async def add_member(
    group_id: UUID,
    data: AddMemberRequest,
    current_user: UserOut = Depends(get_current_user),
    service: GroupService = Depends(get_group_service),
):
    return await service.add_member_to_group(str(group_id), data, current_user.id)


@router.delete("/{group_id}/members/{member_id}")
async def remove_member(
    group_id: UUID,
    member_id: UUID,
    current_user: UserOut = Depends(get_current_user),
    service: GroupService = Depends(get_group_service),
):
    return await service.remove_member_from_group(str(group_id), str(member_id), current_user.id)


@router.delete("/{group_id}")
async def delete_group(
    group_id: UUID,
    current_user: UserOut = Depends(get_current_user),
    service: GroupService = Depends(get_group_service),
):
    return await service.delete_group(str(group_id), current_user.id)

File: frontend/src/lib/api/groups.ts
import { api } from "../http";
import type { PaginatedResponse } from "./types";

export interface Group {
  id: string;
  name: string;
}

export interface GroupMember {
  id: string; // Membership ID
  user_id: string;
  role: string;
  user: {
    id: string;
    name: string;
    email: string;
  };
}

export interface GroupDetail {
  id: string;
  name: string;
  members: GroupMember[];
  user_balance: number;
  total_spent: number;
  total_owed: number;
  total_owe: number;
}

export interface GroupCreate {
  name: string;
  initial_members: string[];
}

export interface AddMemberRequest {
  user_id?: string;
  email?: string;
}

export const GroupsAPI = {
  create(data: GroupCreate) {
    return api.post<Group>("/groups/", data);
  },

  list(search?: string, skip = 0, limit = 20) {
    const params = new URLSearchParams();
    if (search) params.append("search", search);
    params.append("skip", skip.toString());
    params.append("limit", limit.toString());

    const queryString = params.toString();
    return api.get<PaginatedResponse<Group>>(
      `/groups/${queryString ? `?${queryString}` : ""}`
    );
  },

  getDetail(id: string) {
    return api.get<GroupDetail>(`/groups/${id}`);
  },

  addMember(groupId: string, data: AddMemberRequest) {
    return api.post<GroupMember>(`/groups/${groupId}/members`, data);
  },

  removeMember(groupId: string, memberId: string) {
    return api.delete(`/groups/${groupId}/members/${memberId}`);
  },
};
